src/main/java/build/buildfarm/instance/server/NodeInstance.java

```diff
@@ -738,19 +741,57 @@ public ListenableFuture<Digest> fetchBlob(
     return fetchBlobUrls(urls.build(), headers, expectedDigest, requestMetadata);
   }
 
+  /**
+   * Some headers in `allHeaders` are intended ONLY for specific URLs indexes, for example, AUTH
+   * headers that we wouldn't want exposed to all URL (mirrors).
+   *
+   * <p>If a header is prefixed with "<int>:", it is intended for the URL at index <int>. If a
+   * header is not prefixed, it is intended for all URLs. If a header is indexed, it is higher
+   * priority and will override existing headers with the same key.
+   *
+   * @param allHeaders
+   * @param urlIndex
+   * @return
+   */
+  @VisibleForTesting
+  static @Nonnull Map<String, String> getHeadersForUrlIndex(
```

Comment:
Not sure I'm on board with the Nonnull specification - we've been Nullable-ing things that _could_ be null, but everything else is presumed (at least initially) to not be null